File: tokens.py
import inspect
from sys import stderr

ignored_funcs = ["term", "term_any", "term_seq", "_x", "_y", "main", "<module>"]

def token_error(tt, rr, l, i, ex):
    start, end = _y(tt, rr, l, i)
    exs = ", ".join(ex)
    got = _x(tt, rr, l, i)
    if got == "continue":
        print("Errorneous continue (token %d) at (%d:%d)" % (i, start, end), file=stderr)
        return i + 1
    elif got in ["for", "do", "while"]:
        print("Errorneous loop (token %d) at (%d:%d)" % (i, start, end), file=stderr)
        return i + 1
    fmt_args = (i, start, end, exs if len(ex) == 1 else f"one of [{exs}]", got)
    error_text = "Error at token %d (%d:%d): expected %s, got %s" % fmt_args
    print(error_text, file=stderr)
    print("===Token error start===", file=stderr)
    print(", ".join([fr.function.split("_", 1)[-1] for fr in inspect.stack()[1:] if fr.function not in ignored_funcs]), file=stderr)
    print("=== Token error end ===", file=stderr)
    return i + 1

def _x(tt, rr, l, i):
    if i < l:
        return tt[i]
    return "<EOF>"

def _y(tt, rr, l, i):
    if i < l:
        return rr[i]
    # return rr[l-1][1], rr[l-1][1]
    return -1, -1

def term_any(tt, rr, l, i, expected_list):
    if (i < l) and (tt[i] in expected_list):
        i += 1
    else:
        i = token_error(tt, rr, l, i, expected_list)
    return i

def term(tt, rr, l, i, expected):
    return term_any(tt, rr, l, i, [expected])

def term_seq(tt, rr, l, i, seq):
    for expected in seq:
        if isinstance(expected, str):
            i = term(tt, rr, l, i, expected)
        else:
            i = term_any(tt, rr, l, i, expected)
    return i

def parse_classdef(tt, rr, l, i):
    print("class start", i)
    i = term_any(tt, rr, l, i, ["class", "struct"])
    if (i < l) and tt[i] == ":":
        i += 1 # inheritance colon
    if (i < l) and tt[i] == ";":
        i = term(tt, rr, l, i, ";")
    elif (i < l) and tt[i] == "{":
        i = term(tt, rr, l, i, "{")
        # while (i < l) and (tt[i] in [":", "(", ";", "class", "struct"]):
        while (i < l) and tt[i] != "}":
            if tt[i] in [":", ";"]:
                i += 1
            elif tt[i] == "(":
                i = parse_func(tt, rr, l, i)
            elif tt[i] in ["class", "struct"]:
                i = parse_classdef(tt, rr, l, i)
            else:
                i = token_error(tt, rr, l, i, [":", ";", "(", "class", "struct"])
                # i += 1
        i = term(tt, rr, l, i, "}")
    else:
        i = token_error(tt, rr, l, i, ["{", ";"])
    print("class end", i)
    return i

def parse_balanced_expr(tt, rr, l, i):
    if tt[i] == "(":
        i = term(tt, rr, l, i, "(")
        i = parse_balanced_expr(tt, rr, l, i)
        i = term(tt, rr, l, i, ")")
    elif tt[i] == "{":
        i = term(tt, rr, l, i, "{")
        i = parse_balanced_expr(tt, rr, l, i)
        i = term(tt, rr, l, i, "}")
    # inside is optional => no "else error"
    return i

def parse_func(tt, rr, l, i):
    print("func start", i)
    i = term_seq(tt, rr, l, i, ["(", ")", "{"])
    # while (i < l) and (tt[i] in [";", "for", "do", "while"]):
    while (i < l) and (tt[i] != "}"):
        print("func stmt start", i)
        if tt[i] == ";":
            i += 1
        elif tt[i] in ["for", "do", "while"]:
            i = parse_loop(tt, rr, l, i)
        else:
            i = token_error(tt, rr, l, i, [";", "for", "do", "while"])
        print("func stmt end", i)
    i = term(tt, rr, l, i, "}")
    print("func end", i)
    return i

def parse_for(tt, rr, l, i):
    i = term_seq(tt, rr, l, i, ["for", "("])
    if tt[i] == ":":
        # range-based for
        i = term_seq(tt, rr, l, i, [":", ")"])
    elif (i+1 < l) and (tt[i] == ";") and (tt[i+1] == ";"):
        # normal for
        i = term_seq(tt, rr, l, i, [";", ";", ")"])
    else:
        i = token_error(tt, rr, l, i, [";", ":"])
    i = parse_loop_body(tt, rr, l, i)
    return i

def parse_do(tt, rr, l, i):
    i = term(tt, rr, l, i, "do")
    i = parse_loop_body(tt, rr, l, i)
    i = term_seq(tt, rr, l, i, ["while", "("])
    i = parse_balanced_expr(tt, rr, l, i)
    i = term_seq(tt, rr, l, i, [")", ";"])
    return i

def parse_while(tt, rr, l, i):
    i = term_seq(tt, rr, l, i, ["while", "("])
    i = parse_balanced_expr(tt, rr, l, i)
    i = term(tt, rr, l, i, ")")
    i = parse_loop_body(tt, rr, l, i)
    return i

def parse_loop(tt, rr, l, i):
    print(f"loop start ({tt[i]})", i)
    if tt[i] == "for":
        i = parse_for(tt, rr, l, i)
    elif tt[i] == "do":
        i = parse_do(tt, rr, l, i)
    elif tt[i] == "while":
        i = parse_while(tt, rr, l, i)
    else:
        i = token_error(tt, rr, l, i, ["for", "do", "while"])
    print("loop end", i)
    return i

def parse_loop_block(tt, rr, l, i):
    print("loop block start", i)
    i = term(tt, rr, l, i, "{")
    while (i < l) and (tt[i] != "}"):
        print("loop stmt start", i)
        i = parse_loop_stmt(tt, rr, l, i)
        print("loop stmt end", i)
    i = term(tt, rr, l, i, "}")
    print("loop block end", i)
    return i

def parse_loop_stmt(tt, rr, l, i):
    if tt[i] == ";":
        i = term(tt, rr, l, i, ";")
    elif tt[i] == "(":
        i = parse_balanced_expr(tt, rr, l, i)
    elif tt[i] == "continue":
        i = term_seq(tt, rr, l, i, ["continue", ";"])
    else:
        i = token_error(tt, rr, l, i, [";", "(", "continue"])
    return i

def parse_loop_body(tt, rr, l, i):
    print("loop body start", i)
    if (i < l) and (tt[i] == "{"):
        i = parse_loop_block(tt, rr, l, i)
    else:
        i = parse_loop_stmt(tt, rr, l, i)
    print("loop body end", i)
    return i

def parse_namespace(tt, rr, l, i):
    print("namespace start", i)
    i = term_seq(tt, rr, l, i, ["namespace", "{"])
    i = parse_prog(tt, rr, l, i)
    i = term(tt, rr, l, i, "}")
    print("namespace end", i)
    return i

def parse_prog(tt, rr, l, i):
    print("prog start", i)
    while (i < l) and (tt[i] != "}"):
        if tt[i] == ";":
            i += 1
        elif (i+1 < l) and (tt[i] in ["class", "struct"]) and (tt[i+1] in ["{", ";"]):
            i = parse_classdef(tt, rr, l, i)
        elif tt[i] == "namespace":
            i = parse_namespace(tt, rr, l, i)
        elif (i+3 < l) and (tt[i] == "(") and (tt[i+1] == ")") and (tt[i+2] in ["{", ";"]):
            i = parse_func(tt, rr, l, i)
        else:
            i = token_error(tt, rr, l, i, [";", "class", "struct", "namespace", "("])
    print("prog end", i)
    return i

File: src/tokens.c
#include <stdio.h>
#include "types.h"

static inline void print_expected_tokens(FILE *out, kw_t *exl) {
	myfprintf(out, "[%s", keywords[*exl]);
	for (exl++; *exl; exl++) myfprintf(out, ", %s", keywords[*exl]);
	myfprintf(out, "]");
}

static inline size_t token_error(control_t cc, token const *tt, size_t len, size_t i, kw_t *exl) {
	(void)len;
	kw_t t;
	long line, col;
	if (i < len) { t = tt[i].t; line = tt[i].lineno; col = tt[i].colno; }
	else { t = KW_EOF; line = -1; col = -1; }
	myfprintf(cc.ferrs, "Error: (token %zu) expected one of ", i);
	print_expected_tokens(cc.ferrs, exl);
	myfprintf(cc.ferrs, ", but got %s (%zu:%zu)\n", keywords[t], line, col);
	return i+1;
}

static inline size_t term(control_t cc, token const *tt, size_t len, size_t i, kw_t ex) {
	if (i < len && tt[i].t == ex) return i+1;
	return token_error(cc, tt, len, i, (kw_t[]){ex, 0});
}

static inline size_t term_any(control_t cc, token const *tt, size_t len, size_t i, kw_t *exl) {
	if (i < len) {
		for (kw_t t = tt[i].t; *exl; exl++)
			if (t == *exl) return i+1;
	}
	return token_error(cc, tt, len, i, exl);
}

static inline size_t parse_classdef(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_namespace(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_func(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_balanced_expr(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_loop(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_for(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_do(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_while(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_loop_block(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_loop_stmt(control_t cc, token const *tt, size_t len, size_t i);
static inline size_t parse_loop_body(control_t cc, token const *tt, size_t len, size_t i);

static inline size_t parse_classdef(control_t cc, token const *tt, size_t len, size_t i) {
	myfprintf(cc.ftree, "class start %zu\n", i);
	i = term_any(cc, tt, len, i, (kw_t[]){KW_class, KW_struct, 0});
	if (i < len && tt[i].t == ':')
		i++; // inheritance colon
	if (i < len && tt[i].t == ';')
		i++;
	else if (i < len && tt[i].t == '{') {
		i++;
		while (i < len && tt[i].t != '}') {
			if (tt[i].t == ':' || tt[i].t == ';') i++;
			else if (tt[i].t == '(')
				i = parse_func(cc, tt, len, i);
			else if (tt[i].t == KW_class || tt[i].t == KW_struct)
				i = parse_classdef(cc, tt, len, i);
			else
				i = token_error(cc, tt, len, i, (kw_t[]){':', ';', '(', KW_class, KW_struct, 0});
		}
		i = term(cc, tt, len, i, '}');
	}
	else i = token_error(cc, tt, len, i, (kw_t[]){'{', ';', 0});
	myfprintf(cc.ftree, "class end %zu\n", i);
	return i;
}

static inline size_t parse_balanced_expr(control_t cc, token const *tt, size_t len, size_t i) {
	if (tt[i].t == '(') {
		// implied '('
		i = parse_balanced_expr(cc ,tt, len, i+1);
		i = term(cc, tt, len, i, ')');
	}
	else if (tt[i].t == '{') {
		// implied '{'
		i = parse_balanced_expr(cc, tt, len, i+1);
		i = term(cc, tt, len, i, '}');
	}
	// inside is optional => no "else error"
	return i;
}

static inline size_t parse_func(control_t cc, token const *tt, size_t len, size_t i) {
	myfprintf(cc.ftree, "func start %zu\n", i);
	i = term(cc, tt, len, i, '(');
	i = term(cc, tt, len, i, ')');
	i = term(cc, tt, len, i, '{');
	while (i < len && tt[i].t != '}') {
		myfprintf(cc.ftree, "func stmt start %zu\n", i);
		if (tt[i].t == ';') i++;
		else if (tt[i].t == KW_for || tt[i].t == KW_do || tt[i].t == KW_while)
			i = parse_loop(cc, tt, len, i);
		else if (tt[i].t == '(')
			i = parse_balanced_expr(cc, tt, len, i);
		else
			i = token_error(cc, tt, len, i, (kw_t[]){';', KW_for, KW_do, KW_while, 0});
		myfprintf(cc.ftree, "func stmt end %zu\n", i);
	}
	i = term(cc, tt, len, i, '}');
	myfprintf(cc.ftree, "func end %zu\n", i);
	return i;
}

static inline size_t parse_for(control_t cc, token const *tt, size_t len, size_t i) {
	i = term(cc, tt, len, i, KW_for);
	i = term(cc, tt, len, i, '(');
	if (tt[i].t == ':') {
		// range-based for
		// implied ':'
		i = term(cc, tt, len, i+1, ')');
	}
	else if (i+1 < len && tt[i].t == ';' && tt[i+1].t == ';') {
		// normal for
		// implied ';' ';'
		i = term(cc, tt, len, i+2, ')');
	}
	else i = token_error(cc, tt, len, i, (kw_t[]){';', ':', 0});
	i = parse_loop_body(cc, tt, len, i);
	return i;
}

static inline size_t parse_do(control_t cc, token const *tt, size_t len, size_t i) {
	i = term(cc, tt, len, i, KW_do);
	i = parse_loop_body(cc, tt, len, i);
	i = term(cc, tt, len, i, KW_while);
	i = term(cc, tt, len, i, '(');
	i = parse_balanced_expr(cc, tt, len, i);
	i = term(cc, tt, len, i, ')');
	i = term(cc, tt, len, i, ';');
	return i;
}

static inline size_t parse_while(control_t cc, token const *tt, size_t len, size_t i) {
	i = term(cc, tt, len, i, KW_while);
	i = term(cc, tt, len, i, '(');
	i = parse_balanced_expr(cc, tt, len, i);
	i = term(cc, tt, len, i, ')');
	i = parse_loop_body(cc, tt, len, i);
	return i;
}

static inline size_t parse_loop(control_t cc, token const *tt, size_t len, size_t i) {
	myfprintf(cc.ftree, "loop start (%s) %zu\n", keywords[tt[i].t], i);
	if (tt[i].t == KW_for)
		i = parse_for(cc, tt, len, i);
	else if (tt[i].t == KW_do)
		i = parse_do(cc, tt, len, i);
	else if (tt[i].t == KW_while)
		i = parse_while(cc, tt, len, i);
	else
		i = token_error(cc, tt, len, i, (kw_t[]){KW_for, KW_do, KW_while, 0});
	myfprintf(cc.ftree, "loop end %zu\n", i);
	return i;
}

static inline size_t parse_loop_block(control_t cc, token const *tt, size_t len, size_t i) {
	myfprintf(cc.ftree, "loop block start %zu\n", i);
	i = term(cc, tt, len, i, '{');
	while (i < len && tt[i].t != '}') {
		myfprintf(cc.ftree, "loop stmt start %zu\n", i);
		i = parse_loop_stmt(cc, tt, len, i);
		myfprintf(cc.ftree, "loop stmt end %zu\n", i);
	}
	i = term(cc, tt, len, i, '}');
	myfprintf(cc.ftree, "loop block end %zu\n", i);
	return i;
}

static inline size_t parse_loop_stmt(control_t cc, token const *tt, size_t len, size_t i) {
	if (tt[i].t == ';') i++;
	else if (tt[i].t == '(')
		i = parse_balanced_expr(cc, tt, len, i);
	else if (tt[i].t == KW_continue)
		// implied KW_continue
		i = term(cc, tt, len, i+1, ';');
	else i = token_error(cc, tt, len, i, (kw_t[]){';', '(', KW_continue, 0});
	return i;
}

static inline size_t parse_loop_body(control_t cc, token const *tt, size_t len, size_t i) {
    myfprintf(cc.ftree, "loop body start %zu\n", i);
    if (i < len && tt[i].t == '{')
		i = parse_loop_block(cc, tt, len, i);
    else
		i = parse_loop_stmt(cc, tt, len, i);
    myfprintf(cc.ftree, "loop body end %zu\n", i);
    return i;
}

static inline size_t parse_namespace(control_t cc, token const *tt, size_t len, size_t i) {
	myfprintf(cc.ftree, "namespace start %zu\n", i);
	i = term(cc, tt, len, i, KW_namespace);
	i = term(cc, tt, len, i, '{');
	i = parse_prog(cc, tt, len, i);
	i = term(cc, tt, len, i, '}');
	myfprintf(cc.ftree, "namespace end %zu\n", i);
	return i;
}

size_t parse_prog(control_t cc, token const *tt, size_t len, size_t i) {
	myfprintf(cc.ftree, "prog start %zu\n", i);
	while ((i < len) && (tt[i].t != '}')) {
		if (tt[i].t == ';') i++;
		else if (
			i+1 < len
			&& (tt[i].t == KW_class || tt[i].t == KW_class)
			&& (tt[i+1].t == '{' || tt[i+1].t == ';'))
			i = parse_classdef(cc, tt, len, i);
		else if (tt[i].t == KW_namespace)
			i = parse_namespace(cc, tt, len, i);
		else if (
			i+3 < len
			&& tt[i].t == '('
			&& tt[i+1].t == ')'
			&& tt[i+2].t == '{') {
			i = parse_func(cc, tt, len, i);
		}
		else
			i = token_error(cc, tt, len, i, (kw_t[]){';', '(', KW_class, KW_struct, KW_namespace, 0});
	}
	myfprintf(cc.ftree, "prog end %zu\n", i);
	return i;
}
